### PR TITLE
Virtual metadata on item delete

### PR DESCRIPTION
**Pull request by PhilipVis**
This PR continues on the work for #530. It updates the item delete page which opens when clicking on the "Permanently delete" button on the admin edit item page, so that the administrator can inspect the relationships and related virtual metadata of the item that is being deleted. 

At the bottom of the item delete page the relationships of the item that will be deleted are listed. Next to each relationship an info button is available which opens a modal when clicked. The modal lists the virtual metadata of the relationship. 

This PR is still a work in progress, but it should already allow you to start testing these features. Tests and documentation are still incomplete. We’ll push them to this PR in the next few days.

### CODE REVIEW

File: src/app/+item-page/edit-item-page/item-delete/item-delete.component.html [hunk 1]
```diff
@@ -0,0 +1,96 @@
+<div class="container">
```
Comment (paulo-graca) [inline src/app/+item-page/edit-item-page/item-delete/item-delete.component.html]: I think there is room for some improvement regarding this component. Like given the user the real vision of how metadata would look like if he chooses to delete data (which fields and values will be copied after delete the record).

File: src/app/core/data/object-updates/object-updates.service.ts [hunk 2]
```diff
@@ -37,6 +43,10 @@ function filterByUrlAndUUIDFieldStateSelector(url: string, uuid: string): Memoiz
   return createSelector(filterByUrlObjectUpdatesStateSelector(url), (state: ObjectUpdatesEntry) => state.fieldStates[uuid]);
 }
 
+function virtualMetadataSourceSelector(url: string, source: string): MemoizedSelector<CoreState, VirtualMetadataSource> {
```
Comment (paulo-graca) [inline src/app/core/data/object-updates/object-updates.service.ts]: missing typescript docs

File: src/app/+item-page/edit-item-page/item-delete/item-delete.component.spec.ts [hunk 3]
```diff
@@ -1,44 +1,65 @@
...
+import {NotificationsService} from '../../../shared/notifications/notifications.service';
+import {CUSTOM_ELEMENTS_SCHEMA} from '@angular/core';
+import {By} from '@angular/platform-browser';
+import {ItemDeleteComponent} from './item-delete.component';
+import {getItemEditPath} from '../../item-page-routing.module';
+import {createSuccessfulRemoteDataObject} from '../../../shared/testing/utils';
+import {VarDirective} from '../../../shared/utils/var.directive';
+import {ObjectUpdatesService} from '../../../core/data/object-updates/object-updates.service';
+import {RelationshipService} from '../../../core/data/relationship.service';
+import {RelationshipType} from '../../../core/shared/item-relationships/relationship-type.model';
```
Comment (tdonohue) [inline src/app/+item-page/edit-item-page/item-delete/item-delete.component.spec.ts]: There seems to be a lot of unnecessary reformatting in these `import` statements.  Not a big deal, but I *think* we usually have spaces around the curly brackets...and it looks like your IDE removed them

File: src/app/+item-page/edit-item-page/item-delete/item-delete.component.ts [hunk 4]
```diff
@@ -1,31 +1,285 @@
...
+import {NgbModalRef} from '@ng-bootstrap/ng-bootstrap';
+import {combineLatest as observableCombineLatest, combineLatest, Observable} from 'rxjs';
+import {RelationshipType} from '../../../core/shared/item-relationships/relationship-type.model';
+import {VirtualMetadata} from '../virtual-metadata/virtual-metadata.component';
+import {Relationship} from '../../../core/shared/item-relationships/relationship.model';
+import {getRemoteDataPayload, getSucceededRemoteData} from '../../../core/shared/operators';
+import {hasValue, isNotEmpty} from '../../../shared/empty.util';
+import {Item} from '../../../core/shared/item.model';
+import {MetadataValue} from '../../../core/shared/metadata.models';
+import {ViewMode} from '../../../core/shared/view-mode.model';
```
Comment (tdonohue) [inline src/app/+item-page/edit-item-page/item-delete/item-delete.component.ts]: Again, some unnecessary import reformatting.

File: src/app/+item-page/edit-item-page/item-private/item-private.component.spec.ts [hunk 5]
```diff
@@ -59,6 +63,19 @@ describe('ItemPrivateComponent', () => {
...
+        // do nothing
+      },
+    };
+
+    relationshipService = jasmine.createSpyObj('relationshipService',
+      {
+        getItemRelationshipTypesArray: observableOf([]),
+        getItemRelationshipsArray: observableOf([]),
+      }
+    );
```
Comment (tdonohue) [inline src/app/+item-page/edit-item-page/item-private/item-private.component.spec.ts]: I'm not sure I understand why these `objectUpdatesServiceStub` and `relationshipService` blocks of code were added to *all* specs under the `edit-item-page`.  It seems odd to have to initialize these to empty in a bunch of unrelated specs.  I see this same change in `item-public.component.spec.ts`, `item-reinstate.component.spec.ts`, `item-withdraw.component.spec.ts`, `abstract-simple-item-action.component.spec.ts` and `abstract-simple-item-action.component.ts`

Comment (samuelcambien) [reply src/app/+item-page/edit-item-page/item-private/item-private.component.spec.ts]: This was because these services were added in the common AbstractSimpleItemActionComponent constructor. I changed that now, the ItemDeleteComponent related services are now added in the ItemDeleteComponent specific constructor, so only that spec file has to be changed.

**Comment by samuelcambien** [comment]
Rebased all changes onto PR-530 into a single commit.

File: src/app/+item-page/edit-item-page/item-delete/item-delete.component.spec.ts (hunk 3 above)
Comment (samuelcambien) [reply src/app/+item-page/edit-item-page/item-delete/item-delete.component.spec.ts]: I've reverted the imports reformatting

File: src/app/+item-page/edit-item-page/item-delete/item-delete.component.ts (hunk 4 above)
Comment (samuelcambien) [reply src/app/+item-page/edit-item-page/item-delete/item-delete.component.ts]: I've reverted the imports reformatting

File: src/app/core/data/object-updates/object-updates.service.ts (hunk 2 above)
Comment (samuelcambien) [reply src/app/core/data/object-updates/object-updates.service.ts]: I didn't add these because the other selectors don't have typedocs neither.
If necessary I can still add them though.

File: src/app/+item-page/edit-item-page/item-delete/item-delete.component.html (hunk 1 above)
Comment (samuelcambien) [reply src/app/+item-page/edit-item-page/item-delete/item-delete.component.html]: Hi Paulo, thanks for your suggestion, but as you pointed out above, it won't be for this PR.

**Comment by samuelcambien** [comment]
Merged the latest master branch in this PR (Angular 7 upgrade)

**Comment by tdonohue** [comment]
Note: As #530 was just merged, this PR needs rebasing (as it currently includes commits that were already merged in #530).

**Comment by samuelcambien** [comment]
* added a check to skip relationship types without relationships and adapted the tests accordingly
* merged the latest master

File: src/Typescript-codestyle.xml [hunk 6]
```diff
@@ -0,0 +1,50 @@
+<code_scheme name="Default (1)" version="173">
```
Comment (paulo-graca) [inline src/Typescript-codestyle.xml]: is this file required?

Comment (samuelcambien) [reply src/Typescript-codestyle.xml]: Hi Paulo, I've committed this file by accident.
Though it might not be a bad idea perhaps to include a codestyle file in git, it's not related to this PR anyway, so I deleted it again.

**Comment by paulo-graca** [comment]
I'm getting this error:

ItemDeleteComponent.html:11 ERROR TypeError: Cannot read property 'length' of null
    at Object.eval [as updateDirectives] (ItemDeleteComponent.html:11)
    at Object.debugUpdateDirectives [as updateDirectives] (client.js:74202)
    at checkAndUpdateView (client.js:73598)
    at callViewAction (client.js:73839)
    at execEmbeddedViewsAction (client.js:73802)
    at checkAndUpdateView (client.js:73599)
    at callViewAction (client.js:73839)
    at execComponentViewsAction (client.js:73781)
    at checkAndUpdateView (client.js:73604)
    at callViewAction (client.js:73839)

"<div *ngIf="types.length > 0" class="mb-4">"

When after saving metadata and switching into the "info status" tab

**Comment by samuelcambien** [comment]
> I'm getting this error:
> 
> ItemDeleteComponent.html:11 ERROR TypeError: Cannot read property 'length' of null
> at Object.eval [as updateDirectives] (ItemDeleteComponent.html:11)
> at Object.debugUpdateDirectives [as updateDirectives] (client.js:74202)
> at checkAndUpdateView (client.js:73598)
> at callViewAction (client.js:73839)
> at execEmbeddedViewsAction (client.js:73802)
> at checkAndUpdateView (client.js:73599)
> at callViewAction (client.js:73839)
> at execComponentViewsAction (client.js:73781)
> at checkAndUpdateView (client.js:73604)
> at callViewAction (client.js:73839)
> 
> "<div *ngIf="types.length > 0" class="mb-4">"
> 
> When after saving metadata and switching into the "info status" tab

Hi Paulo, this extra check should resolve this issue:
https://github.com/DSpace/dspace-angular/pull/533/commits
